openwillis/measures/text/util/speech/coherence.py

```diff
@@ -0,0 +1,557 @@
+            similarity_matrix = cosine_similarity(utterances_embeddings) if sentence_encoder else None
+
+            # Initialize coherence lists
+            sentence_tangeniality1_list, sentence_tangeniality2_list = [], []
+            perplexity_list, perplexity_5_list, perplexity_11_list, perplexity_15_list = [], [], [], []
+            turn_to_turn_tangeniality_list = []
+            for i, row in utterances_filtered.iterrows():
+                current_speaker = row[measures['speaker_label']]
+
+                if current_speaker != speaker_label:
+                    continue
+                elif len(row[measures['words_texts']]) < min_coherence_turn_length:
+                    sentence_tangeniality1_list.append(np.nan)
+                    sentence_tangeniality2_list.append(np.nan)
+                    perplexity_list.append(np.nan)
+                    perplexity_5_list.append(np.nan)
+                    perplexity_11_list.append(np.nan)
+                    perplexity_15_list.append(np.nan)
+                    turn_to_turn_tangeniality_list.append(np.nan)
+                    continue
+
+                phrases_texts = row[measures['phrases_texts']]
+                utterance_text = row[measures['utterance_text']]
+                
+                sentence_tangeniality1, sentence_tangeniality2, perplexity, perplexity_5, perplexity_11, perplexity_15 = calculate_phrase_tangeniality(
+                    phrases_texts, utterance_text, sentence_encoder, bert, tokenizer
+                )
+                
+                sentence_tangeniality1_list.append(sentence_tangeniality1)
+                sentence_tangeniality2_list.append(sentence_tangeniality2)
+                perplexity_list.append(perplexity)
+                perplexity_5_list.append(perplexity_5)
+                perplexity_11_list.append(perplexity_11)
+                perplexity_15_list.append(perplexity_15)
+
+                if i == 0 or len(utterances_filtered.iloc[i - 1][measures['words_texts']]) < min_coherence_turn_length or not sentence_encoder:
+                    turn_to_turn_tangeniality_list.append(np.nan)
+                else:
+                    turn_to_turn_tangeniality_list.append(similarity_matrix[i, i - 1])
+
+            turn_df[measures['sentence_tangeniality1']] = sentence_tangeniality1_list
+            turn_df[measures['sentence_tangeniality2']] = sentence_tangeniality2_list
+            turn_df[measures['perplexity']] = perplexity_list
+            turn_df[measures['perplexity_5']] = perplexity_5_list
+            turn_df[measures['perplexity_11']] = perplexity_11_list
+            turn_df[measures['perplexity_15']] = perplexity_15_list
+            turn_df[measures['turn_to_turn_tangeniality']] = turn_to_turn_tangeniality_list
+
+
+            for measure in ['sentence_tangeniality1', 'sentence_tangeniality2', 'perplexity', 'perplexity_5', 'perplexity_11', 'perplexity_15', 'turn_to_turn_tangeniality']:
```

Comment:
just checking: do we need to handle anything here?

Reply:
addressed but no need for further exception handling